src/bilevelJuMPExtension.jl

```diff
@@ -0,0 +1,170 @@
+    local model_fixed = remove_local_var(model_local,semi_locals)
+    local s = repr(eval(model_fixed))
+    local p = r"\n.*return"
+    local blm = match(r"\b.*(?==.*Bilevel)",s).match
+    local vars_temp = [split(x.match," ")[3] for x in eachmatch(r"(?<=@variable).*",s)]
+    # println(vars_temp)
+    local vars = [Meta.parse(replace(x,r"(?<=\[).*(?=\])|\]|\["=>"")) for x in vars_temp] #get the vars defined in the model
+    # println(vars)
+    if length(expression_local) == 1
+        local locals = []
+    else
+        local cref,locals = remain_cref(expression_local[1])
+    end
+    local ctr, = remain_opt_var(model_local,expression_local[end],vars,semi_locals,locals)#modify the contraint expression, always the last argument
+    ctr = string(ctr)
+    cref = string(cref)
+    if length(expression_local) == 1
+        local constraint = "@constraint(Upper($blm),$(ctr))"
+    elseif length(expression_local) == 2
+        local constraint = "@constraint(Upper($blm),$(cref),$(ctr))"
+    end
+    local m = replace(s,p=>(pstr->"$constraint\n"*pstr))
+    local model_new = Meta.parse(m)
+    # @info("constraint ***$constraint*** added")
+    eval(model_new)
+    # println(m)
+    end
+end
+macro add_lower_cons(model,expression...)
+    model_string = string(model)
+    expression_string = [string(e) for e in expression]
+    return quote
+    local model_local = Meta.parse($model_string)
+    local expression_local = [Meta.parse(x) for x in $expression_string]
+    @assert length($expression_string) <= 2
+    @assert length($expression_string) >= 1
+    local semi_locals = Base.@locals # get the local vars within the loop
+    local model_fixed = remove_local_var(model_local,semi_locals)
+    local s = repr(eval(model_fixed))
+    local p = r"\n.*return"
+    local blm = match(r"\b.*(?==.*Bilevel)",s).match
+    local vars_temp = [split(x.match," ")[3] for x in eachmatch(r"(?<=@variable).*",s)]
+    # println(vars_temp)
+    local vars = [Meta.parse(replace(x,r"(?<=\[).*(?=\])|\]|\["=>"")) for x in vars_temp] #get the vars defined in the model
+    # println(vars)
+    if length(expression_local) == 1
+        local locals = []
+    else
+        local cref,locals = remain_cref(expression_local[1])
+    end
+    local ctr, = remain_opt_var(model_local,expression_local[end],vars,semi_locals,locals)#modify the contraint expression, always the last argument
+    ctr = string(ctr)
+    cref = string(cref)
+    if length(expression_local) == 1
+        local constraint = "@constraint(Lower($blm),$(ctr))"
+    elseif length(expression_local) == 2
+        local constraint = "@constraint(Lower($blm),$(cref),$(ctr))"
+    end
+    local m = replace(s,p=>(pstr->"$constraint\n"*pstr))
+    local model_new = Meta.parse(m)
+    # @info("constraint ***$constraint*** added")
+    eval(model_new)
+    # println(m)
+    end
+end
+function remain_opt_var(model,expression,varlist,semi_locals,local_var=[])
```

Comment:
It would be good to add docstrings to the new functions.
Also comments in macros are ver important, they tend to be obscure to most users.